### DISCUSSION TITLE
Have a problem when working with the Form Client Component in Nextjs 13.4

### DISCUSSION THREAD

**Discussion by pingpongdoctor** [vercel/next.js#49414] [Help]
### Summary

I am practicing with Nextjs 13.4 App Router and I have a problem when working with the Form Client Component. I have a todo list page and in this page I have a TodoList component and a AddTodo component. AddTodo component is a client component where I can submit a new todo to the server with a form. When I try to type something in the input box, the page get stuck, meaning that I can not implement any user interactivity. My page get stuck with no errors when I run npm run dev or npm run build. Please help me.

- Code in page.tsx:

import { Metadata } from "next";
import AddTodo from "./components/AddTodo";
import TodoList from "./components/TodoList";

export const metadata: Metadata = {
  title: "Todo list",
};

//SET REVALIDATE 0 TO MAKE THE PAGE RENDERED AS A SSR PAGE (NO DATA CACHING)
export const revalidate = 0;

export default function Todos() {
  return (
    <>
      {/* @ts-expect-error Async Server Component */}
      <AddTodo />
      {/* @ts-expect-error Async Server Component */}
      <TodoList />
    </>
  );
}

- Code in AddToDo Client Component where the problem takes place:

"use client";

import handleAddTodo from "@/lib/handleAddTodo";
import { ChangeEvent, FormEvent } from "react";
import { useState } from "react";
import { useRouter } from "next/navigation";

export default async function AddTodo() {
  const [name, setName] = useState<string>("");

  const handleChangeName = function (e: ChangeEvent<HTMLInputElement>) {
    setName(e.target.value);
  };
  const router = useRouter();
  const handleAdd = async function (e: FormEvent<HTMLFormElement>) {
    e.preventDefault();
    if (name) {
      await handleAddTodo(name);
      router.refresh();
    }
  };

  return (
    <form onSubmit={handleAdd}>
      <input onChange={handleChangeName} type="text" />
      <button>Add Todo</button>
    </form>
  );
}

- Code of handleAddTodo Function that is created in lib file:

export default async function handleAddTodo(name: string): Promise<void> {
  const res = await fetch("http://localhost:8000/todos", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ name }),
  });

  if (!res.ok) {
    throw new Error("add todos error");
  }

  return;
}

- Code of Todo List Server Component (provided in case you need to refer to):

import getTodos from "@/lib/getTodos";
import Todo from "./Todo";

export default async function TodoList() {
  const todoData: Promise<Todo[]> = getTodos();
  const todos = await todoData;

  return (
    <ul>
      {todos?.length > 0 &&
        todos.map((todo) => <Todo key={todo.id} todoObj={todo} />)}
      <h1>todo</h1>
    </ul>
  );
}

- Code of Todo Client Component (provided in case you need to refer to):
"use client";

import handlDeleteTodo from "../../../lib/handleDeleteTodo";
import { useState, useTransition } from "react";
import { useRouter } from "next/navigation";

interface Props {
  todoObj: {
    id: number;
    name: string;
  };
}

export default function Todo({ todoObj: { id, name } }: Props) {
  const [isFetching, setIsFetching] = useState<boolean>(false);
  const [isPending, startTransition] = useTransition();
  const isMutating = isFetching || isPending;
  const router = useRouter();

  const handleDelete = async function (id: number) {
    setIsFetching(true);

    await handlDeleteTodo(id);

    setIsFetching(false);

    startTransition(() => {
      router.refresh();
    });
  };

  return (
    <li style={{ opacity: !isMutating ? 1 : 0.7 }}>
      <p>
        Todo thing with id:{id} and name:{name}
      </p>
      <button
        onClick={() => {
          handleDelete(id);
        }}
      >
        Delete
      </button>
    </li>
  );
}

**Comment by KevinCotoCarrera** [comment]
The same problem as you, mate

**Reply by icyJoseph** [reply]
Could you provide some code? Preferably with code fencing, https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks#fenced-code-blocks, so that it is a bit more readable.

**Reply by KevinCotoCarrera** [reply]
I'm just using MUI input components in a 'use client' page. I actually have some external components would be a bit trick to share code. When I click whatever input in order to type the page just get stuck

**Reply by KevinCotoCarrera** [reply]
`'use client'
import axios from 'axios'
import { TextField, Grid, Box, Button, FormGroup, SelectChangeEvent } from "@mui/material"
import { useRouter } from 'next/navigation';
import { useFormik} from 'formik'
import { SignupSchema } from './schema';
import GridWithTextField from '../components/Form/GridWithTextField';
import { useEffect, useState } from 'react';
import SelectWithLabel from '../components/Form/SelectWithLabel';
import { GetServerSideProps } from 'next';    
import useSWR from 'swr';

const serverURL = process.env.NEXT_PUBLIC_SERVER_URL

interface Country {
  key: React.Key, 
  value: string
  alpha_code2: string
}
interface State {
  key: React.Key, 
  value: string
}
interface City {
  key: React.Key, 
  value: string
}
// async function getCountries () {
//   const res = await fetch(`${serverURL}/nomenclatures/countries`);
//   return res.json()
//   // const serverURL = process.env.NEXT_PUBLIC_SERVER_URL
//   // console.log(serverURL)
//   // const res = await axios.get(`${serverURL}/nomenclatures/countries`, {params: {orderDir: "ASC"}})
//   // if(res.data.length == 0){
//   //   throw new Error('Failed to get Countries Data')
//   // }
//   // const parsedRes = JSON.parse(JSON.stringify(res.data))
//   // const countries : Country[] = parsedRes.data
//   // return countries

// }
const fetcher = (url: RequestInfo | URL) => fetch(url).then(r => r.json())
export default async function SignupForm(){
    const router = useRouter();
    const countries = useSWR(`${serverURL}/nomenclatures/countries?orderDir=ASC` , fetcher)
    // const countries: Country[] = await getCountries()
    const [states, setStates] = useState<State[]>([])
    const [cities, setCities] = useState<City[]>([])
    const [countryKey, setCountryKey] = useState('233');
    const [stateKey, setStateKey] = useState('')
    const [cityKey, setCityKey] = useState('')

    const handleCountryChange = (e: SelectChangeEvent<typeof countryKey>) => {
      setCountryKey(e.target.value)
    }
    const handleStateChange = (e: SelectChangeEvent<typeof stateKey>) => {
      setStateKey(e.target.value)
    }
    const handleCityChange = (e: SelectChangeEvent<typeof cityKey>) => {
      setCityKey(e.target.value)
    }
    // const countries = getCountries()
    // useEffect(() =>{
    //   (async() =>{

    //   })
    // }, [countries])

    const {values,errors, handleBlur, handleChange, handleSubmit} = useFormik({  
        initialValues: {
         
            phone: "",
            firstName: "",
            middleName: "",
            lastName: "",
            secondLastName: "",
            dateOfBirth: "",
            gender: "",
            email: "",
            acceptSms: false,       
            streetAddress: "",
            countryId: "",
            stateId: "",
            cityId: "",
            "postalCode": ""
        },
        onSubmit:async (values: any) => {
          console.log('onSubmit')
          const firstName = values.firstName
    
          const signupData = {
            firstName: firstName,
          }
          axios({
            method: 'POST',
            url: `${serverURL}/user/register`,
            data: signupData
          })
            .then(function (res: any) {
               console.log(res)
               alert('Successfully signed up!')
               router.push('/login')
    
            })
            .catch(function (res: any) {
               console.log(res)
          });
        },
        validationSchema: SignupSchema,
      });
    
    return(
        <Grid container
        direction={'row'}
        justifyItems={'center'}
        justifyContent={'center'}
        sx={{
            width: '100vw',
            height: '100vh',
            placeItems: 'center',
            backgroundColor: 'green'
        }}
        >
            <Grid 
            item
            container
            direction={'row'}
            justifyContent={'center'}
            sx={{
              width: '400px',
              height: '80vh',
              paddingX: '8px',
              paddingTop: '32px',
              backgroundColor: 'white',
            }}
            component={'form'}
            onSubmit={handleSubmit}
            
            >   
                  <GridWithTextField
                  id="firstName"
                  name="firstName"
                  type="firstName"
                  label="Nombre"
                  onChange={handleChange}
                  onBlur={handleBlur}
                  value={values.firstName}
                  error={errors.firstName}                  
                  />
                  <GridWithTextField
                  id="middleName"
                  name="middleName"
                  type="middleName"
                  label="Segundo Nombre"
                  onChange={handleChange}
                  onBlur={handleBlur}
                  value={values.middleName}
                  error={errors.middleName}                  
                  />
                  <GridWithTextField
                  id="lastName"
                  name="lastName"
                  type="lastName"
                  label="Apellido"
                  onChange={handleChange}
                  onBlur={handleBlur}
                  value={values.lastName}
                  error={errors.lastName}                  
                  />
                  <GridWithTextField
                  id="secondLastName"
                  name="secondLastName"
                  type="secondLastName"
                  label="Segundo Apellido"
                  onChange={handleChange}
                  onBlur={handleBlur}
                  value={values.secondLastName}
                  error={errors.secondLastName}                  
                  />
                  <GridWithTextField
                  id="streetAddress"
                  name="streetAddress"
                  type="streetAddress"
                  label="Dirección"
                  onChange={handleChange}
                  onBlur={handleBlur}
                  value={values.streetAddress}
                  error={errors.streetAddress}                  
                  />
                  <GridWithTextField
                  id="postalCode"
                  name="postalCode"
                  type="postalCode"
                  label="Código Postal"
                  onChange={handleChange}
                  onBlur={handleBlur}
                  value={values.postalCode}
                  error={errors.postalCode}                  
                  />
                  {countries.isLoading ? 'Loading Countries...' :
                    <SelectWithLabel
                    id='countries'
                    name='select-countries'
                    label='Selecciona un País'
                    data={countries.data}
                    value={countryKey}
                    onChange={handleCountryChange}
                    
                    />}
                    <Button variant="contained" type='submit'>Continuar</Button>
            </Grid>
        </Grid>
    )
}
 
`

**Comment by KevinCotoCarrera** [comment]
I changed to pages structure folder in next 13.4.1 and it worked as usual. The problem is 
undoubtedly that Page Router isn't stable yet.

**Comment by a95tbirdsc** [comment]
I believe I have the same problem using the new APP directory.  I have a simple input field where a user can type in someones name.  Like typing in J...Jo...Joh...John.  This search needs to be passed to a Server Component that queries the API for J, Jo, Joh and John.  Using "export default async function SearchResults ({ query })" and then using async/await with the fetch API just ends up calling the SearchResults function until the end of time.

Has anyone figured out how to have a Search field and then pass that information to a Server Component with an async/await fetch API?  Can it even be done with the new APP directory?

**Comment by pingpongdoctor** [comment]
Hey buddies, I eventually figured it out. In my code, I have the form client component that is created with async function. I only need to convert the async function to normal function and it works well. I guess Nextjs automatically finds async components as server components. This also means that if you create an async component, the "use client" in your client component will be deactivated and the component will work as a server component. The form component must be a client component since it needs user interactivity (React hydration).

In general, when attempting to type something in an input element and no changes are seen, it may be due to the fact that the input element is rendered on the server since the "use client" is ignored when we create the client component with async function.

**Reply by a95tbirdsc** [reply]
Do you have any examples I can look at?  As I mentioned, I have an input field (page.js has 'use client') and whatever you type in the input field is sent to SearchResults with a "<div>{query}</div>" statement showing the results.  That works.  Change SearchResults to be "export default **async** function SearchResults" that uses "const response = **await fetch**('https://.....)" and fetch will continue to call the URL until the end of time.

I can't seem to figure out how to have an Input Field ('use client') that calls an ASYNC/AWAIT fetch API within a Server Component work and display the results without it calling the fetch API until the end of time.  From what I can see, React/Next JS just can't do it.

**Reply by icyJoseph** [reply]
You can't have so called, client hooks (useState, useEffect), which give interactivity, within Server Components.

As soon as a component is using async/await, it is out of the question for client hooks. 

In fact, it is interactivity that's somewhat curved down. You cannot even pass a callback from a Server Component down to a client component. Server actions are an API to bring interactivity into the mix.

You can't attach an event listener on a button click, etc, within a server component. 

For now, if a component must use useState, or other client hooks, it has to be a Client Component.

**Reply by a95tbirdsc** [reply]
So how to you get it to work?

This does work but there's **no async or await fetch** involved.

SearchText (page.js)
`
'use client'

import { useCallback, useState } from 'react'
import utilStyles from '/styles/utils.module.css';
import { faMagnifyingGlass, faPeopleGroup, faBuildingUser, faArrowsRotate } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import SearchResults from '../search-results/page.js';

export default function SearchText() {
  const [query, setQuery] = useState('');
  
  const onChange = useCallback((event) => {
    const query = event.target.value;
    setQuery(query);
  }, []);

  console.log({query});
  return (
      
    <>
      
      <form id='EmployeeSearchText'>
        <div className="container-fluid p-3">
          <div id='Row1' className={`row ${utilStyles.divPadding}`}>
            <div id='Row1Column1' className="col">
              <div id='SearchText-container' className={utilStyles.searchBar}>
                <FontAwesomeIcon icon={faMagnifyingGlass} className={utilStyles.icon} />
                <input id='SearchText' type="text" onChange={onChange} className={`form-control ${utilStyles.inputField} ${utilStyles.inputPadding}`} name="SearchText" value={query} placeholder="Search Last Name; First Name; Campus; Deparment....." />
              </div>
            </div>
          </div>
          <div className='text-center'>
            <div id='Row2' className={`row ${utilStyles.divPadding}`}>
              <div id='Row2Column1' className="col">
                <div id='Employees-container'>
                  <button id='EmployeeListing' className={utilStyles.btnYellow}>Employees&nbsp;&nbsp;<FontAwesomeIcon icon={faPeopleGroup} /></button>
                </div>
              </div>
              <div id='Row2Column2' className="col">
                <div id='DepartmentListing-container'>
                  <button id='DepartmentListing' className={utilStyles.btnYellow} type="button">Department Listing&nbsp;&nbsp;<FontAwesomeIcon icon={faBuildingUser} /></button>
                </div>
              </div>
              <div id='Row2Column3' className="col">
                <div id='SearchTextReset-container'>
                  <button id='SearchTextReset' className={utilStyles.btnGrey} type="button">Reset&nbsp;&nbsp;<FontAwesomeIcon icon={faArrowsRotate} /></button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </form>

      <SearchResults query={query} />
      
    </>
  
  )
}

`

SearchResults (page.js)

`
import utilStyles from '/styles/utils.module.css';
import { faPhone, faEnvelope } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';

export default function SearchResults( {query} ) {
  return (
    
    <div className="container">
      <div className="row">{query}</div>
    </div>
  
  );
  
}

`

Once I add "export default async function SearchResults" that uses "const response = await fetch('https://.....)" and fetch will continue to call the URL until the end of time.

So, show me some code that works having a simple Input Field that calls a Server Component that uses ASYNC AWAIT Fetch.  I mean, do the components have to be assembled in a certain way.  Meaning, <Client Component><Server Component><Server Component><Client Component> or some other combination.  One has to call one that calls another that then calls another that goes back to the first component.

**Reply by icyJoseph** [reply]
Could you please, try to use https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks - I can't see where your code starts or ends.

> This is what I see:

![Screenshot 2023-05-09 at 18 15 20](https://github.com/vercel/next.js/assets/21013447/b1a46623-589f-4355-a209-e84d11c50b9e)

Could you also just list out, step 1, 2, 3, etc, what is it you are trying to do? that's more helpful.

From what I can see right now, your client component, passes `query` to `SearchResults`, and that's the ultimate problem. 

- https://nextjs.org/docs/getting-started/react-essentials#unsupported-pattern

Server Components are a network boundary of sorts, and React puts a slot where they should go. 

As part of the rendering process of a client component though, you cannot include a Server Component. Instead, composition is used.

- https://nextjs.org/docs/getting-started/react-essentials#recommended-pattern-passing-server-components-to-client-components-as-props

What you could do in your case, is use URL queries. In this answer I give such an example: https://github.com/vercel/next.js/discussions/47227#discussioncomment-5342500

I am not sure how server actions would help here either. 

Do remember that passing that `query` from client to server, implies a network round trip... so it is not trivial to pull off.

It is not like, this is not part of the React Server Components spec though: take this experimental framework for instance https://github.com/dai-shi/wakuwork, fully explained here: https://www.youtube.com/watch?v=ciQdQkT0bZk

**Reply by a95tbirdsc** [reply]
icyJoseph,

Are you saying there's no way possible to pass user Input to a Server Component?  You can't have an Input Field that passes that information to a Server Component that uses that information in the API Fetch call.

Let's take the Google Search.  There is no way to implement a Google Search, using React/Next JS, that lets you type in some value and have some Server Component use ASYNC AWAIT Fetch returning that result set back to a Client Component to display the results.  A simple Google Search can't be built using React/Next JS.

**Reply by icyJoseph** [reply]
I don't know what you are on about, but of course it can be built. A google search url ends with `/search?q=hello` that's a piece of cake.

What is a bit more tricky is, having a client component that holds state, and passing that onto a server component, and get that one to do its magic. 

I've managed to do that with Suspense boundaries, and SWR, but not server component. We are still in the early days of this technology, only the app route is marked as stable. 

I guess [Server Actions](https://nextjs.org/docs/app/building-your-application/data-fetching/server-actions) or some other pattern might step up to take this use case.

**Reply by a95tbirdsc** [reply]
icyJoseph,

I clicked the Add Code button but it appears you need triple ticks.

Pretty simple.  A Google Search but in my case a search for Faculty/Staff/Students.

Step 1. A web page with a simple Input Field (Just like Google Search).
Step 2. User inputs some text (partial name, department, college.....etc) (Just like you do on Google Search).
Step 3. The User's Input is passed to a function that uses ASYNC AWAIT Fetch to return back a result set.
Step 4. That result set is displayed under the simple Input Field (Just like when doing a Google Search and the results are shown below).

Proposal:  How do I create a page that functions just like Google Search using an Input Field that uses React/Next JS.  I already have the API Endpoint.

An actual code sample would help me understand what I need to do.  Your example doesn't even need to work but show me how the Client and Server Components are assembled and the flow and I can insert my API Endpoint and other html components in.  I just need something to go on so I can finish this application.

![image](https://github.com/vercel/next.js/assets/132099076/86678415-edcf-4c70-a5e6-83c89543c9d0)

**Reply by a95tbirdsc** [reply]
I see.  So, in reality, what I am wanting to do just doesn't work at the moment.
I looked at some of your links you provided and I think I have tried every combination trying to get something to work.

Pretty sure I have tried this and it just calls the FETCH until the end of time.

```
// This pattern works:
// You can pass a Server Component as a child or prop of a
// Client Component.
import ExampleClientComponent from './example-client-component';
import ExampleServerComponent from './example-server-component';
 
// Pages in Next.js are Server Components by default
export default function Page() {
  return (
    <ExampleClientComponent>
      <ExampleServerComponent />
    </ExampleClientComponent>
  );
}

```

This works.  As you can see I typed in something in the input field and that text is displayed below.

SearchText (Input Field that passes the User's input to SearchResults as query)

```
'use client'

import { useCallback, useState } from 'react'
import utilStyles from '/styles/utils.module.css';
import { faMagnifyingGlass, faPeopleGroup, faBuildingUser, faArrowsRotate } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import SearchResults from '../search-results/page.js';

export default function SearchText() {
  const [query, setQuery] = useState('');
  
  const onChange = useCallback((event) => {
    const query = event.target.value;
    setQuery(query);
  }, []);

  console.log({query});
  return (
      
    <>
      
      <form id='EmployeeSearchText'>
        <div className="container-fluid p-3">
          <div id='Row1' className={`row ${utilStyles.divPadding}`}>
            <div id='Row1Column1' className="col">
              <div id='SearchText-container' className={utilStyles.searchBar}>
                <FontAwesomeIcon icon={faMagnifyingGlass} className={utilStyles.icon} />
                <input id='SearchText' type="text" onChange={onChange} className={`form-control ${utilStyles.inputField} ${utilStyles.inputPadding}`} name="SearchText" value={query} placeholder="Search Last Name; First Name; Campus; Deparment....." />
              </div>
            </div>
          </div>
          <div className='text-center'>
            <div id='Row2' className={`row ${utilStyles.divPadding}`}>
              <div id='Row2Column1' className="col">
                <div id='Employees-container'>
                  <button id='EmployeeListing' className={utilStyles.btnYellow}>Employees&nbsp;&nbsp;<FontAwesomeIcon icon={faPeopleGroup} /></button>
                </div>
              </div>
              <div id='Row2Column2' className="col">
                <div id='DepartmentListing-container'>
                  <button id='DepartmentListing' className={utilStyles.btnYellow} type="button">Department Listing&nbsp;&nbsp;<FontAwesomeIcon icon={faBuildingUser} /></button>
                </div>
              </div>
              <div id='Row2Column3' className="col">
                <div id='SearchTextReset-container'>
                  <button id='SearchTextReset' className={utilStyles.btnGrey} type="button">Reset&nbsp;&nbsp;<FontAwesomeIcon icon={faArrowsRotate} /></button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </form>

      <SearchResults query={query} />
      
    </>
  
  )
}

```


SearchResults

```
import utilStyles from '/styles/utils.module.css';
import { faPhone, faEnvelope } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';

export default function SearchResults( {query} ) {
  return (
    
    <div className="container">
      <div className="row">{query}</div>
    </div>
  
  );
  
}

```

![image](https://github.com/vercel/next.js/assets/132099076/b3c0b8ab-92a4-4089-a62a-6a671bcec612)


Now, I just need to get that query information to a Fetch and return the result set.

**Comment by pingpongdoctor** [comment]
I guess the problem is that you are trying to use a server component inside a client component and this is not accepted unless you use  the children prop. In you code, you have a client page that include form and the imported searchResult page. The form works well since it is a client component, but the imported searchResult page is seen as a server component by default in Nextjs (page.tsx is always server component). This would then lead to the conflict in which a server component (searchResult page) is imported inside a client component (searchText).

Solution is that you can import the client component to the page.tsx instead of import the page.tsx to the client component.

**Reply by a95tbirdsc** [reply]
I don't quite follow you so here's all my code.  This is my latest version.  

Just tell me what changes are needed in the files that will make this work.  You can add any additional files you want.  Just show/tell me what they are and how they fit in.

This version works but never shows any results from the Fetch API.  It appears to be called when you refresh the page but no results are shown.

app/page.js

```

import SearchHeader from './search-header/page.js';
import SearchText from './search-text/page.js';
import SearchResults from './search-results/page.js';

export default function Home() {
  return (
    <main>
      <div>
        <SearchHeader />
        <SearchText>
          <SearchResults />
        </SearchText>
      </div>
    </main>
  )
}

```

app/search-header/page.js

```

import utilStyles from '/styles/utils.module.css';
import { faArrowDown } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';

export default function Header() {
  return (
      
    <>
      
      <h2 className={utilStyles.pageHeader}>Employee Directory</h2>
      <div id="Tips-container">
        <span id="Tips">
          <div>
            <div id="searchPanel" className={utilStyles.searchPanel}>
              <div className={utilStyles.card}>
                <div className={utilStyles.cardBody}>
                  <h5 className={utilStyles.cardTitle}>Search Tips</h5>
                  <p className={utilStyles.cardText}></p>
                  <ul>
                    <li>Employees can be searched across multiple criteria.  By entering in a few characters, Employees can be matched by Last Name, First Name, Campus, Department, Jobtitle, Phone or Email.  A search can be narrowed down by entering in more characters or widened by only entering in one or two characters.</li>
                    <li>Selecting the "Employee" button will provide an alphabetically employee list sorted by last name.</li>
                    <li>Selecting the "Department Listing" button will provide an alphabetically list of employees within each department.</li>
                  </ul>
                  <p></p>
                </div>
              </div>
            </div>
            <div className={`${utilStyles.searchFlapPrimary} ${utilStyles.searchFlap} ${utilStyles.searchFlapDefault}`}>Click for Searching Tips.&nbsp;&nbsp;<FontAwesomeIcon icon={faArrowDown} className={utilStyles.searchIcon} /></div>
          </div>
        </span>
      </div>
    </>
  
  )
}

```

app/search-text/page.js

```
'use client';

import { useCallback, useState } from 'react';
import utilStyles from '/styles/utils.module.css';
import { faMagnifyingGlass, faPeopleGroup, faBuildingUser, faArrowsRotate } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import SearchResults from '../search-results/page.js';

export default function SearchText({ children }) {
  const [query, setQuery] = useState('');
  
  const onChange = useCallback((event) => {
    const query = event.target.value;
    setQuery(query);
  }, []);

  console.log({query});
  return (
      
    <>
      
      <form id='EmployeeSearchText'>
        <div className="container-fluid p-3">
          <div id='Row1' className={`row ${utilStyles.divPadding}`}>
            <div id='Row1Column1' className="col">
              <div id='SearchText-container' className={utilStyles.searchBar}>
                <FontAwesomeIcon icon={faMagnifyingGlass} className={utilStyles.icon} />
                <input id='SearchText' type="text" onChange={onChange} className={`form-control ${utilStyles.inputField} ${utilStyles.inputPadding}`} name="SearchText" value={query} placeholder="Search Last Name; First Name; Campus; Deparment....." />
              </div>
            </div>
          </div>
          <div className='text-center'>
            <div id='Row2' className={`row ${utilStyles.divPadding}`}>
              <div id='Row2Column1' className="col">
                <div id='Employees-container'>
                  <button id='EmployeeListing' className={utilStyles.btnYellow}>Employees&nbsp;&nbsp;<FontAwesomeIcon icon={faPeopleGroup} /></button>
                </div>
              </div>
              <div id='Row2Column2' className="col">
                <div id='DepartmentListing-container'>
                  <button id='DepartmentListing' className={utilStyles.btnYellow} type="button">Department Listing&nbsp;&nbsp;<FontAwesomeIcon icon={faBuildingUser} /></button>
                </div>
              </div>
              <div id='Row2Column3' className="col">
                <div id='SearchTextReset-container'>
                  <button id='SearchTextReset' className={utilStyles.btnGrey} type="button">Reset&nbsp;&nbsp;<FontAwesomeIcon icon={faArrowsRotate} /></button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </form>

    </>
  
  )
}

```

app/search-results/page.js

```

import utilStyles from '/styles/utils.module.css';
import { faPhone, faEnvelope } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';

export default async function SearchResults( {query} ) {

  async function getEmployees() {
    const res = await fetch('https://xxxxx');
    
    const directory = await res.json();
    return directory;
  }
  
  const employees = await getEmployees();

  const emp = employees.map((employee, index, array) => {
    
    return (
        
      <div className="col col-12 col-sm-6 col-xl-4 mb-4">
        <div className="col-12 mb-4">
          <div className={`card ${utilStyles.cardBody}`}>
            <div className="card-body">
              <div className={`${utilStyles.ribbonLight} ${utilStyles.ribbonBottomRight}`}><span>{employee.CAMPUS}</span></div>
              <h3 className="card-title card-name">{employee.NAME}</h3>
              <h5 className={`${utilStyles.cardSubtitle} ${utilStyles.cardJobtitleNew} mb-2`}>{employee.JOBTITLE}</h5>
              <p className={`${utilStyles.cardText} ${utilStyles.cardDepartment}`}>{employee.DEPARTMENT}</p>
              <div>
                <p className={utilStyles.cardTextBottom}><FontAwesomeIcon icon={faPhone} className={utilStyles.cardTextIcons} /><span className={utilStyles.cardPhoneEmail}>W:</span>{employee.PHONE}</p>
                <p className={utilStyles.cardTextBottom}><FontAwesomeIcon icon={faEnvelope} className={utilStyles.cardTextIcons} /><span className={utilStyles.cardPhoneEmail}>W:</span></p>
              </div>
            </div>
          </div>
        </div>
      </div>
    
    )
      
  });
  
  return (
    
    <div className="container">
      <div className="row">{emp}</div>
    </div>
       
  );
}

```

**Reply by pingpongdoctor** [reply]
You should create a components folder in the searchResult directory and create a SearchTextComponent in the new components folder. Then, you can import the searchText component (remember to use "use client" in this component) to the searchResult page. The problem is that you can not use "use client" in a page.js to convert this page to a client component because a page.js is always seen as a server component. You need to use components folder as I mentioned above. In short, Do not use "use client" in page.ts.

**Reply by icyJoseph** [reply]
> because a page.js is always seen as a server component. You need to use components folder as I mentioned above. In short, Do not use "use client" in page.tsx

That's incorrect. A page can be a client component.

```jsx
"use client";

export default function Home() {
  return (
    <main>
      <h1>Hello I am a client component</h1>
    </main>
  );
}
```

The problem is that a client component cannot directly import a server component, because server components have to be isolated from the client modules.

In such cases, you need a third component to act as orchestrator, that sits above the client component and the server component we want to put together. Since you can't have anything higher up than pages (other than layout, or template), the page itself has to be turned into a server component.

Please read the docs, https://nextjs.org/docs/getting-started/react-essentials#recommended-pattern-passing-server-components-to-client-components-as-props, it is all explained there with great detail:

```jsx
// This page cannot opt into 'use client', because it is used as an orchestrator
// to put together ExampleClientComponent and ExampleServerComponent 


// This pattern works:
// You can pass a Server Component as a child or prop of a
// Client Component.
import ExampleClientComponent from './example-client-component';
import ExampleServerComponent from './example-server-component';
 
// Pages in Next.js are Server Components by default
export default function Page() {
  return (
    <ExampleClientComponent>
      <ExampleServerComponent />
    </ExampleClientComponent>
  );
}
```

**Reply by a95tbirdsc** [reply]
I'm confused.  It would appear I'm doing exactly what icyJoseph is showing.  Am I doing something else wrong?

Is this incorrect?


app/page.js

```
import SearchHeader from './search-header/page.js';
import SearchText from './search-text/page.js';
import SearchResults from './search-results/page.js';

export default function Home() {
  return (
    <main>
      <div>
        <SearchHeader />
        <SearchText>
          <SearchResults />
        </SearchText>
      </div>
    </main>
  )
}

```

**Reply by icyJoseph** [reply]
Mmm you are importing pages into pages? you don't have to call these `page.js` because that makes them into routes. Could you post  your tree structure?

```console
tree -I node_modules -L 3 src/app
src/app
├── blog
│   ├── [slug]
│   │   └── page.tsx
│   └── page.tsx
├── layout.tsx
├── not-found.tsx
├── page.tsx
└── views
    └── route.ts

3 directories, 6 files
```

The way you have it, I could, for example, go to `app/search-header`, and that'll be a route.

**Reply by a95tbirdsc** [reply]
I get the following information using the command you have shown:

tree -I node_modules -L 3 src/app

Too many parameters - node_modules


I then entered the following comman:
tree -I

Folder PATH listing for volume Windows
Volume serial number is 5C4D-A599
C:\REACT\EMPLOYEE-DIRECTORY\-I
Invalid path - \REACT\EMPLOYEE-DIRECTORY\-I
No subfolders exist

**Reply by icyJoseph** [reply]
Ah sorry I assumed Linux OS :( a screenshot would be alright then. Unless you figure out how to print tree dirs in Windows. My bad.

**Reply by a95tbirdsc** [reply]
Hope this helps.

c:\react\employee-directory\

![image](https://github.com/vercel/next.js/assets/132099076/8d52da9a-dd31-463b-8d44-e08693e6664e)

c:\react\employee-directory\app\

![image](https://github.com/vercel/next.js/assets/132099076/5f826135-fd6a-4cb2-a7d4-745d6d01e61c)

**Reply by icyJoseph** [reply]
Right, why would you import pages into pages though? Also are you using types module on your project? Asking because you have file extensions on your imports.

Last but not least, what errors are you seeing? Screenshots help with that too.

**Reply by a95tbirdsc** [reply]
This might be better for you.  Super long, though.

[tree.txt](https://github.com/vercel/next.js/files/11444156/tree.txt)

**Reply by a95tbirdsc** [reply]
It sounds like I don't have a clue of what I'm doing.  I thought that's how you got this to work.

So, you don't do this?

```
import SearchHeader from './search-header/page.js';
import SearchText from './search-text/page.js';
import SearchResults from './search-results/page.js';

export default function Home() {
  return (
    <main>
      <div>
        <SearchHeader />
        <SearchText>
          <SearchResults />
        </SearchText>
      </div>
    </main>
  )
}

```

No clue on Types Module.  Don't even know what that is.  Are we not supposed to put ".js" on our page file?

I have provided all my code in an earlier post.  The folder with each page.js file.  Can you show me what I need to do?  It sounds like what I'm doing is all wrong.

No errors but no results show up.

I can get results (all faculty/staff) to display if I do the following but then I don't know how to pass the User's Input to SearchResults.

```
import SearchHeader from './search-header/page.js';
import SearchText from './search-text/page.js';
import SearchResults from './search-results/page.js';

export default function Home() {
  return (
    <main>
      <div>
        <SearchHeader />
        <SearchText />
        <SearchResults />
      </div>
    </main>
  )
}


```

**Reply by a95tbirdsc** [reply]
Ok, I got this to work with displaying all the records using the "[Recommended Pattern: Passing Server Components to Client Components as Props](https://nextjs.org/docs/getting-started/react-essentials#recommended-pattern-passing-server-components-to-client-components-as-props)" method.  

How do I get the Server Component (SearchResults) to accept the Query (what the user is typing in)?

```
export async function SearchResults( {query} ) {

```

This doesn't work.

```
fetch(https://xxxxxx?searchinput={query});

fetch(https://xxxxxx?searchinput=${query});

```

Any ideas?
